ext/bcmath/libbcmath/src/recmul.c

```diff
@@ -316,6 +346,225 @@ static void bc_standard_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc
+
+	/*
+	 * Carry may occur during subtraction. If we perform multiplication while a carry is occurring,
+	 * BC_VECTOR may overflow sooner than expected, so we calculate the carry.
+	 */
+	if (*buf >= BC_VECTOR_BOUNDARY_NUM && *buf <= BC_VECTOR_MAX_HALF) {
+		*carry = 1;
+		*buf -= BC_VECTOR_BOUNDARY_NUM;
+	} else if ((BC_VECTOR) 0 - *buf >= BC_VECTOR_BOUNDARY_NUM && *buf > BC_VECTOR_MAX_HALF) {
+		*carry = -1;
+		*buf += BC_VECTOR_BOUNDARY_NUM;
+	} else {
+		*carry = 0;
+	}
+}
+
+/*
+ * When a subtraction results in a carry that cannot fit in the buffer, add the result to mid.
+ * Carry is always 1 or -1, so can just add/subtract the buffer as is.
+ */
+static inline void bc_karatsuba_mul_add_buf_to_mid(BC_VECTOR *mid, const BC_VECTOR *buf, BC_VECTOR carry, size_t half_size)
+{
+	if (EXPECTED(carry == 0)) {
+		return;
+	}
+
+	if (carry == 1) {
+		for (size_t i = 0; i < half_size; i++) {
+			mid[i + half_size] += buf[i];
+		}
+	} else {
+		for (size_t i = 0; i < half_size; i++) {
+			mid[i + half_size] -= buf[i];
+		}
+	}
+}
+
+/*
+ * The Karatsuba algorithm uses recursive divide-and-conquer computation. This function is
+ * a process that is called recursively.
+ */
+static BC_VECTOR *bc_karatsuba_mul_recursive(BC_VECTOR *n1, BC_VECTOR *n2, size_t calc_arr_size)
+{
+	BC_VECTOR *ret = safe_emalloc(calc_arr_size * 2, sizeof(BC_VECTOR), 0);
+
+	if (calc_arr_size == 2) {
+		BC_VECTOR low = n1[0] * n2[0];
+		BC_VECTOR high = n1[1] * n2[1];
+		ret[0] = low;
+		ret[1] = low + high - (n1[1] - n1[0]) * (n2[1] - n2[0]);
+		ret[2] = high;
+		ret[3] = 0;
+		return ret;
+	}
+
+	size_t i;
+	size_t half_size = calc_arr_size / 2;
+
+	/* low */
+	BC_VECTOR *low = bc_karatsuba_mul_recursive(n1, n2, half_size);
+	/* high */
+	BC_VECTOR *high = bc_karatsuba_mul_recursive(n1 + half_size, n2 + half_size, half_size);
+
+	/* prepare mid */
+	BC_VECTOR *n1_buf = safe_emalloc(calc_arr_size, sizeof(BC_VECTOR), 0);
+	BC_VECTOR *n2_buf = n1_buf + half_size;
+	BC_VECTOR n1_buf_carry = 0;
+	BC_VECTOR n2_buf_carry = 0;
+	for (i = 0; i < half_size; i++) {
+		/*
+		 * Calculation of carry during subtraction is also performed at the same time.
+		 * Carries that cannot fit into the buffer are recorded in n1_buf_carry and n2_buf_carry.
+		 */
+		bc_karatsuba_mul_buf_calc_for_mid(&n1_buf[i], n1[i + half_size], n1[i], &n1_buf_carry);
+		bc_karatsuba_mul_buf_calc_for_mid(&n2_buf[i], n2[i + half_size], n2[i], &n2_buf_carry);
+	}
+
+	/* mid */
+	BC_VECTOR *mid = bc_karatsuba_mul_recursive(n1_buf, n2_buf, half_size);
+
+	/* If n1_buf has a carry, add n2_buf. The opposite is true when n2_buf has a carry. */
+	bc_karatsuba_mul_add_buf_to_mid(mid, n1_buf, n2_buf_carry, half_size);
+	bc_karatsuba_mul_add_buf_to_mid(mid, n2_buf, n1_buf_carry, half_size);
+
+	/* Add to ret */
+	for (i = 0; i < calc_arr_size; i++) {
+		ret[i] = low[i];
+		ret[i + calc_arr_size] = high[i];
+	}
+	for (i = 0; i < calc_arr_size; i++) {
+		ret[i + half_size] += low[i] + high[i] - mid[i];
+	}
```

Comment:
This is because ret is not initialized to 0, so need to assign low and high to it once, and then add mid, etc. If initialize it to 0 beforehand, only need to loop once.